### PR TITLE
Fix `between` in select query

### LINKED ISSUE #1196: Can not use `between` in the select list:
**Describe the bug**
I can not write a query that uses  `between` in the select list of a sql query

**To Reproduce**

```shell
echo "1" > /tmp/foo.csv
cargo run -p datafusion-cli
```

```sql
CREATE EXTERNAL TABLE foo(x int)
STORED AS CSV
LOCATION '/tmp/foo.csv';
```

```sql
> select 1 between 5 and 10;
NotImplemented("Create name does not support logical expression Utf8(NULL) BETWEEN Int64(5) AND Utf8(NULL)")

> select 1 between 5 and 10 from foo;
NotImplemented("Create name does not support logical expression Int64(1) BETWEEN Int64(5) AND Int64(10)")
```

**Expected behavior**
I expect a single row `false` to be returned

**Workaround**
Use an alias:
```
> select 1 between 5 and 10 as my_expr;
+---------+
| my_expr |
+---------+
| false   |
+---------+
1 row in set. Query took 0.002 seconds.
> select 1 between 5 and 10 as my_expr from foo;
+---------+
| my_expr |
+---------+
| false   |
+---------+
1 row in set. Query took 0.011 seconds.
```

**Additional context**

### PR DESCRIPTION
**Pull request by capkurmagati**
# Which issue does this PR close?

Closes #1196.

1. Added match arms for `Expr::Between`, `Expr::Sort`, and `Expr::Wildcard` and removed default match  in `expr.rs`.
2. Added match arm for `Expr::Between` in `planner.rs`.

```
echo "1" > /tmp/foo.csv
cargo run -p datafusion-cli
```
```
CREATE EXTERNAL TABLE foo(x int)
STORED AS CSV
LOCATION '/tmp/foo.csv';
```

```
>  select 1 between 5 and 10;
+----------------+
| Boolean(false) |
+----------------+
| false          |
+----------------+
1 row in set. Query took 0.003 seconds.

> select 1 between 5 and 10 from foo;
+----------------+
| Boolean(false) |
+----------------+
| false          |
+----------------+
1 row in set. Query took 0.011 seconds.

> select x between 5 and 10 from foo;
+--------------------------------------+
| foo.x BETWEEN Int64(5) AND Int64(10) |
+--------------------------------------+
| false                                |
+--------------------------------------+
1 row in set. Query took 0.012 seconds.

> explain select x between 5 and 10 from foo;
+---------------+---------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                |
+---------------+---------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: #foo.x BETWEEN Int64(5) AND Int64(10)                                                                   |
|               |   TableScan: foo projection=Some([0])                                                                               |
| physical_plan | ProjectionExec: expr=[CAST(x@0 AS Int64) >= 5 AND CAST(x@0 AS Int64) <= 10 as foo.x BETWEEN Int64(5) AND Int64(10)] |
|               |   RepartitionExec: partitioning=RoundRobinBatch(8)                                                                  |
|               |     CsvExec: files=[/tmp/foo.csv], has_header=false, batch_size=8192, limit=None                                    |
|               |                                                                                                                     |
+---------------+---------------------------------------------------------------------------------------------------------------------+
2 rows in set. Query took 0.005 seconds.


```

### CODE REVIEW

**Comment by capkurmagati** [comment]
@Jimexist Sure. Since I'm new to the project let drop some newbie questions.
1. `planner.rs` seems have integration tests using `arrow-datafusion/testing`. Is that the test you want to add?
2. I couldn't find tests for `create_name(e: &Expr, input_schema: &DFSchema) -> Result<String>` . Do you also want to add some tests for that?

**Comment by xudong963** [comment]
> * `planner.rs` seems have integration tests using `arrow-datafusion/testing`. Is that the test you want to add?
> * I couldn't find tests for `create_name(e: &Expr, input_schema: &DFSchema) -> Result<String>` . Do you also want to add some tests for that?

I think you can add unit tests firstly in `datafusion/tests/sql.rs` where has tests about other exprs.

**Comment by alamb** [comment]
> I think you can add unit tests firstly in datafusion/tests/sql.rs where has tests about other exprs.

 I agree with @xudong963  that a test in `sql.rs` would be sufficient for this PR. Thank you @capkurmagati

**Comment by capkurmagati** [comment]
@xudong963 @alamb Thanks for the advice. I added some tests for the expr. PTAL.
I will rebase the branch after reflecting your reviews.

File: datafusion/tests/sql.rs [hunk 1]
```diff
@@ -5305,3 +5305,57 @@ async fn case_with_bool_type_result() -> Result<()> {
...
+        "+--------------------------------------------------------------------+",
+        "| true                                                               |",
+        "| true                                                               |",
+        "| false                                                              |",
+        "| false                                                              |",
+        "+--------------------------------------------------------------------+",
+    ];
+    assert_batches_eq!(expected, &actual);
+
+    let sql = "EXPLAIN SELECT c1 BETWEEN 2 AND 3 FROM test";
```
Comment (capkurmagati) [inline datafusion/tests/sql.rs:5344]: Maybe it's better to to test `NOT BETWEEN` here.

File: datafusion/tests/sql.rs [hunk 2]
```diff
@@ -5305,3 +5305,57 @@ async fn case_with_bool_type_result() -> Result<()> {
...
+        "+---------------+",
+    ];
+    assert_batches_eq!(expected, &actual);
+
+    let input = Int64Array::from(vec![1, 2, 3, 4]);
+    let batch = RecordBatch::try_from_iter(vec![("c1", Arc::new(input) as _)]).unwrap();
+    let table = MemTable::try_new(batch.schema(), vec![vec![batch]])?;
+    ctx.register_table("test", Arc::new(table))?;
+
+    let sql = "SELECT abs(c1) BETWEEN 0 AND LoG(c1 * 100 ) FROM test";
```
Comment (capkurmagati) [inline datafusion/tests/sql.rs:5329]: `LoG` is on purpose to test the exprs are correctly converted recursively.

File: datafusion/src/logical_plan/expr.rs [hunk 3]
```diff
@@ -1935,10 +1935,32 @@ fn create_name(e: &Expr, input_schema: &DFSchema) -> Result<String> {
...
+                Ok(format!("{} NOT BETWEEN {} AND {}", expr, low, high))
+            } else {
+                Ok(format!("{} BETWEEN {} AND {}", expr, low, high))
+            }
+        }
+        Expr::Sort {
+            expr,
+            asc: _,
+            nulls_first: _,
+        } => Err(DataFusionError::Internal(format!(
```
Comment (xudong963) [inline datafusion/src/logical_plan/expr.rs:1957]: We can directly add `Err(DataFusionError::Internal(
            "Create name does not support sort expr".to_string()`)

File: datafusion/src/physical_plan/planner.rs [hunk 4]
```diff
@@ -176,6 +176,21 @@ fn create_physical_name(e: &Expr, is_first_expr: bool) -> Result<String> {
...
+            let expr = create_physical_name(expr, false)?;
+            let low = create_physical_name(low, false)?;
+            let high = create_physical_name(high, false)?;
+            if *negated {
+                Ok(format!("{} NOT BETWEEN {} AND {}", expr, low, high))
+            } else {
+                Ok(format!("{} BETWEEN {} AND {}", expr, low, high))
+            }
+        }
         other => Err(DataFusionError::NotImplemented(format!(
```
Comment (xudong963) [inline datafusion/src/physical_plan/planner.rs:194]: Keep the same as above, add `Sort` and `Wildcard` arms.

File: datafusion/src/logical_plan/expr.rs [hunk 5]
```diff
@@ -1935,10 +1935,32 @@ fn create_name(e: &Expr, input_schema: &DFSchema) -> Result<String> {
                 Ok(format!("{} IN ({:?})", expr, list))
             }
         }
-        other => Err(DataFusionError::NotImplemented(format!(
+        Expr::Between {
```
Comment (xudong963) [inline datafusion/src/logical_plan/expr.rs:1938]: 👍

**Comment by xudong963** [comment]
The remaining LGTM, wait for @alamb to review.

File: datafusion/tests/sql.rs (hunk 1 above)
Comment (alamb) [reply datafusion/tests/sql.rs:5344]: I think this one is fine -- thank you @capkurmagati

**Review by alamb — COMMENTED** [review]
Thank you so much @capkurmagati  -- the tests look great.

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
There appears to be a conflict in sql.rs. I have merged up from main to resolve it

**Comment by alamb** [comment]
The test was failing due to the executor having 2 cores, but the output reflected 8 cores -- updated in https://github.com/apache/arrow-datafusion/pull/1202/commits/8eff3dec094ed0293dd30adb35324617993fb6c7

**Comment by capkurmagati** [comment]
@xudong963 Removed the default match arm and added match arms for `Expr::Sort`, and `Expr::Wildcard` in `create_physical_name` just like `create_name`. PTAL.
@alamb Thanks so much for resolving code conflict and fixing the test. Much appreciated.

File: datafusion/src/logical_plan/expr.rs [hunk 6]
```diff
@@ -1972,10 +1972,31 @@ fn create_name(e: &Expr, input_schema: &DFSchema) -> Result<String> {
...
+            let expr = create_name(expr, input_schema)?;
+            let low = create_name(low, input_schema)?;
+            let high = create_name(high, input_schema)?;
+            if *negated {
+                Ok(format!("{} NOT BETWEEN {} AND {}", expr, low, high))
+            } else {
+                Ok(format!("{} BETWEEN {} AND {}", expr, low, high))
+            }
+        }
+        Expr::Sort {
```
Comment (xudong963) [inline datafusion/src/logical_plan/expr.rs:1990]: You can directly use `Expr::Sort{ .. }`

File: datafusion/src/physical_plan/planner.rs [hunk 7]
```diff
@@ -182,10 +182,31 @@ fn create_physical_name(e: &Expr, is_first_expr: bool) -> Result<String> {
...
+            let expr = create_physical_name(expr, false)?;
+            let low = create_physical_name(low, false)?;
+            let high = create_physical_name(high, false)?;
+            if *negated {
+                Ok(format!("{} NOT BETWEEN {} AND {}", expr, low, high))
+            } else {
+                Ok(format!("{} BETWEEN {} AND {}", expr, low, high))
+            }
+        }
+        Expr::Sort {
```
Comment (xudong963) [inline datafusion/src/physical_plan/planner.rs:200]: ditto

File: datafusion/src/logical_plan/expr.rs (hunk 6 above)
Comment (capkurmagati) [reply datafusion/src/logical_plan/expr.rs]: Thanks for the advice. Fixed.

**Review by alamb — APPROVED** [review]
This is looking great -- thank you @capkurmagati  -- I'll plan to merge this once the CI checks pass

**Comment by alamb** [comment]
Thanks again @capkurmagati !

**Pull request merged**